lib/core/src/sync/mod.rs

```diff
@@ -1,49 +1,211 @@
+            if record.version.floor() > CURRENT_SCHEMA_VERSION.floor() {
+                failed_records.push(record);
+                continue;
+            }
+
+            let decrypted_record =
+                match DecryptedRecord::try_from_record(self.signer.clone(), record) {
+                    Ok(record) => record,
+                    Err(err) => {
+                        warn!("Could not decrypt record: {err:?}");
+                        continue;
+                    }
+                };
+
+            updatable_records.push(decrypted_record)
+        }
+
+        (updatable_records, failed_records)
+    }
+}
 
 #[async_trait]
-impl SyncModule for BreezSyncModule {
+impl SyncService for BreezSyncService {
     async fn connect(&self) -> Result<()> {
-        unimplemented!()
+        let mut client = self.client.lock().await;
+        *client = Some(SyncerClient::connect(self.connect_url.clone()).await?);
+        Ok(())
     }
 
-    async fn apply_changes(&self, changes: &[Record]) -> Result<()> {
-        unimplemented!()
+    async fn listen(self: Arc<Self>) -> Result<()> {
+        let Some(ref mut client) = *self.client.lock().await else {
+            return Err(anyhow!(
+                "Cannot run `get_changes_since`: client not connected"
+            ));
+        };
+
+        let request = ListenChangesRequest::new(utils::now(), self.signer.clone())
+            .map_err(|err| anyhow!("Could not sign ListenChangesRequest: {err:?}"))?;
+        let mut stream = client.listen_changes(request).await?.into_inner();
+
+        let cloned = self.clone();
+        tokio::spawn(async move {
+            while let Some(message) = stream.next().await {
+                match message {
+                    Ok(record) => {
+                        if let Err(err) = cloned.apply_changes(&[record]) {
+                            warn!("Could not apply incoming changes: {err:?}")
+                        };
+                    }
+                    Err(err) => warn!("An error occured while listening for records: {err:?}"),
+                }
+            }
+        });
+
+        Ok(())
     }
 
-    async fn get_changes_since(&self, record_index: u64) -> Result<Vec<Record>> {
-        unimplemented!()
+    fn apply_changes(&self, records: &[Record]) -> Result<()> {
+        let (updatable_records, failed_records) = self.collect_records(records);
+
+        // We persist records which we cannot update (> CURRENT_SCHEMA_VERSION)
+        for record in failed_records {
+            self.persister.insert_record(record)?;
+        }
+
+        // We apply records which we can update (<= CURRENT_SCHEMA_VERSION)
+        for record in updatable_records {
+            if let Err(err) = self.persister.apply_record(record) {
+                warn!("Could not apply record changes: {err:?}");
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_changes_since(&self, from_id: i64) -> Result<Vec<Record>> {
```

Comment:
Who should call get_changes_sync? If it is internal in the sync service perhaps we don't need it in the trait?

Reply:
See https://github.com/breez/breez-sdk-liquid/pull/509#discussion_r1804157040